kv/replica_slice.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package kv
+
+import (
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+// replicaInfo extends the Replica structure with the associated node
+// descriptor.
+type replicaInfo struct {
+	proto.Replica
+	NodeDesc proto.NodeDescriptor
+}
+
+func (i *replicaInfo) attrs() []string {
+	return i.NodeDesc.GetAttrs().Attrs
+}
+
+// A replicaSlice is a slice of Replicas.
+type replicaSlice []replicaInfo
+
+// newReplicaSlice creates a replicaSlice from the replicas listed in the range
+// descriptor and using gossip to lookup node descriptors. Replicas on nodes
+// that are not gossipped are omitted from the result.
+func newReplicaSlice(gossip *gossip.Gossip, desc *proto.RangeDescriptor) replicaSlice {
+	replicas := make(replicaSlice, 0, len(desc.Replicas))
+	for _, r := range desc.Replicas {
+		nd, err := gossip.GetNodeDescriptor(r.NodeID)
+		if err != nil {
+			if log.V(1) {
+				log.Infof("node %d is not gossiped: %v", r.NodeID, err)
+			}
+			continue
+		}
+		replicas = append(replicas, replicaInfo{
+			Replica:  r,
+			NodeDesc: nd,
+		})
+	}
+	return replicas
+}
+
+// Swap interchanges the replicas stored at the given indices.
+func (rs replicaSlice) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+// FindReplica returns the replica which matches the specified store
+// ID. If no replica matches, (-1, nil) is returned.
+func (rs replicaSlice) FindReplica(storeID proto.StoreID) (int, *proto.Replica) {
```

Comment:
It might have been used by `RangeDescriptor.FindReplica` which now duplicates this code. I'll remove in a subsequent change.